core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -349,87 +355,39 @@ public Context(
-                        message += ": no rules used significant amount of time";
-                    }
-                    else {
-                        message += ": Top rules: " + topRulesByTime.stream()
-                                .map(entry -> format("%s: %s", entry.getKey(), entry.getValue()))
-                                .collect(joining(",\n\t\t", "{\n\t\t", " }"));
-                    }
+                    message += ": Top rules: " + topRulesByTime.stream()
+                            .map(ruleStats -> format(
+                                    "%s: %s ms, %s invocations, %s applications",
+                                    ruleStats.getRule(),
+                                    ruleStats.getTotalTime(),
+                                    ruleStats.getInvocations(),
+                                    ruleStats.getApplications()))
+                            .collect(joining(",\n\t\t", "{\n\t\t", " }"));
                 }
                 throw new TrinoException(OPTIMIZER_TIMEOUT, message);
             }
         }
 
-        public void recordRuleInvocation(Rule<?> rule, boolean invoked, boolean applied, long elapsedNanos)
+        void recordRuleInvocation(Rule<?> rule, boolean invoked, boolean applied, long elapsedNanos)
         {
-            ruleStats.computeIfAbsent(rule, ignored -> new RuleInvocationStats())
-                    .recordRuleInvocation(invoked, applied, elapsedNanos);
-        }
-
-        private static class RuleInvocationStats
-        {
-            private long invocationCount;
-            private long applicationCount;
-            private long elapsedTimeNanos;
-
-            void recordRuleInvocation(boolean invoked, boolean applied, long elapsedNanos)
-            {
-                if (invoked) {
-                    invocationCount++;
-                }
-                if (applied) {
-                    applicationCount++;
-                }
-                elapsedTimeNanos += elapsedNanos;
-            }
-
-            public long getInvocationCount()
-            {
-                return invocationCount;
-            }
-
-            public long getApplicationCount()
-            {
-                return applicationCount;
-            }
-
-            public long getElapsedTimeNanos()
-            {
-                return elapsedTimeNanos;
-            }
-
-            private long getElapsedTimeMillis()
-            {
-                return NANOSECONDS.toMillis(elapsedTimeNanos);
-            }
-
-            @Override
-            public String toString()
-            {
-                long elapsedMillis = getElapsedTimeMillis();
-                return format("%s ms, %s invocations, %s applications", elapsedMillis, invocationCount, applicationCount);
-            }
+            // record statistics per query
+            planOptimizersStatsCollector.recordRule(rule, invoked, applied, elapsedNanos);
+            // record statistics per optimizer
```

Comment:
It will be better to move these code comments where these fields are defined for the class